File: tests/cpp/cartridge_manifest_test.cpp
#include "common/serializer/binencoder.h"
#include "rompack/manifest.h"
#include "spec/bmsx/memory_map.h"

#include <stdexcept>
#include <string_view>
#include <utility>

namespace {

bmsx::BinObject hardwareManifest(bmsx::BinArray hardware) {
	bmsx::BinObject root;
	root.emplace("hardware", bmsx::BinValue(std::move(hardware)));
	return root;
}

bmsx::BinValue ramDevice(bmsx::i64 bytes) {
	bmsx::BinObject device;
	device.emplace("type", "ram");
	device.emplace("bytes", bytes);
	return bmsx::BinValue(std::move(device));
}

bmsx::BinValue romDevice() {
	bmsx::BinObject device;
	device.emplace("type", "rom");
	return bmsx::BinValue(std::move(device));
}

bmsx::BinValue mailboxDevice() {
	bmsx::BinObject device;
	device.emplace("type", "mailbox");
	return bmsx::BinValue(std::move(device));
}

bmsx::CartManifest decodeManifest(bmsx::BinObject root) {
	const std::vector<bmsx::u8> encoded = bmsx::encodeBinary(
		bmsx::BinValue(std::move(root))
	);
	bmsx::CartRomHeader header;
	header.manifestLength = static_cast<bmsx::u32>(encoded.size());
	return bmsx::decodeCartManifest(encoded, header);
}

void requireRejected(bmsx::BinObject root, std::string_view label) {
	try {
		(void)decodeManifest(std::move(root));
	} catch (const std::runtime_error&) {
		return;
	}
	throw std::runtime_error(std::string(label) + " was accepted");
}

} // namespace

int main() {
	bmsx::BinArray devices;
	devices.emplace_back(romDevice());
	devices.emplace_back(ramDevice(256));
	devices.emplace_back(mailboxDevice());
	bmsx::BinObject valid = hardwareManifest(std::move(devices));
	valid.emplace("title", "Expansion test");
	const bmsx::CartManifest manifest = decodeManifest(std::move(valid));
	if (manifest.title != "Expansion test"
		|| manifest.hardware.size() != 3u
		|| !std::holds_alternative<bmsx::CartridgeRomDeviceConfig>(
			manifest.hardware[0])
		|| !std::holds_alternative<bmsx::CartridgeRamDeviceConfig>(
			manifest.hardware[1])
		|| std::get<bmsx::CartridgeRamDeviceConfig>(manifest.hardware[1]).bytes != 256u
		|| !std::holds_alternative<bmsx::CartridgeMailboxDeviceConfig>(
			manifest.hardware[2])) {
		throw std::runtime_error("Concrete cartridge hardware did not decode in order");
	}

	requireRejected({}, "Manifest without hardware");

	bmsx::BinObject retired;
	retired.emplace("hardware", bmsx::BinArray{});
	retired.emplace("cartridge", bmsx::BinObject{});
	requireRejected(std::move(retired), "Retired cartridge board schema");

	bmsx::BinObject unknownType;
	unknownType.emplace("type", "math");
	bmsx::BinArray unknownDevices;
	unknownDevices.emplace_back(std::move(unknownType));
	requireRejected(
		hardwareManifest(std::move(unknownDevices)),
		"Unknown hardware device type"
	);

	bmsx::BinObject mailboxWithField;
	mailboxWithField.emplace("type", "mailbox");
	mailboxWithField.emplace("bytes", bmsx::i64{1});
	bmsx::BinArray unknownFieldDevices;
	unknownFieldDevices.emplace_back(std::move(mailboxWithField));
	requireRejected(
		hardwareManifest(std::move(unknownFieldDevices)),
		"Unknown hardware device field"
	);

	bmsx::BinArray duplicateRom;
	duplicateRom.emplace_back(romDevice());
	duplicateRom.emplace_back(romDevice());
	requireRejected(
		hardwareManifest(std::move(duplicateRom)),
		"Duplicate ROM device"
	);

	bmsx::BinObject romWithField;
	romWithField.emplace("type", "rom");
	romWithField.emplace("bytes", bmsx::i64{1});
	bmsx::BinArray romFieldDevices;
	romFieldDevices.emplace_back(std::move(romWithField));
	requireRejected(
		hardwareManifest(std::move(romFieldDevices)),
		"Unknown ROM device field"
	);

	bmsx::BinArray duplicateRam;
	duplicateRam.emplace_back(ramDevice(1));
	duplicateRam.emplace_back(ramDevice(2));
	requireRejected(
		hardwareManifest(std::move(duplicateRam)),
		"Duplicate RAM device"
	);

	bmsx::BinArray duplicateMailbox;
	duplicateMailbox.emplace_back(mailboxDevice());
	duplicateMailbox.emplace_back(mailboxDevice());
	requireRejected(
		hardwareManifest(std::move(duplicateMailbox)),
		"Duplicate mailbox device"
	);

	bmsx::BinArray zeroRam;
	zeroRam.emplace_back(ramDevice(0));
	requireRejected(hardwareManifest(std::move(zeroRam)), "Zero-byte RAM device");

	bmsx::BinArray oversizedRam;
	oversizedRam.emplace_back(ramDevice(static_cast<bmsx::i64>(bmsx::CART_RAM_SIZE) + 1));
	requireRejected(
		hardwareManifest(std::move(oversizedRam)),
		"Oversized RAM device"
	);

	bmsx::BinObject floatingRamDevice;
	floatingRamDevice.emplace("type", "ram");
	floatingRamDevice.emplace("bytes", bmsx::f32{256.0f});
	bmsx::BinArray floatingRam;
	floatingRam.emplace_back(std::move(floatingRamDevice));
	requireRejected(
		hardwareManifest(std::move(floatingRam)),
		"Floating-tag RAM byte count"
	);
}

File: tests/rompacker/cartridge_manifest.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';

import { encodeBinary } from '../../machine/ts/common/serializer/binencoder';
import {
	decodeCartManifest,
	parseCartManifest,
} from '../../machine/ts/rompack/manifest';
import type { CartRomHeader } from '../../machine/ts/rompack/format';
import { CART_RAM_SIZE } from '../../machine/ts/spec/bmsx/memory_map';
import { CART_ROM_HEADER_SIZE } from '../../machine/ts/spec/bmsx/rom_package';
import { writeCartRomHeader } from '../../toolchain/ts/rompack/header_encode';

const EMPTY_HEADER: CartRomHeader = {
	headerSize: CART_ROM_HEADER_SIZE,
	manifestOffset: 0,
	manifestLength: 0,
	tocOffset: 0,
	tocLength: 0,
	dataOffset: 0,
	dataLength: 0,
	blua32ImageOffset: 0,
	blua32ImageByteCount: 0,
	blua32StartupFunctionAddress: 0,
	blua32IrqFunctionAddress: 0,
	blua32ExceptionFunctionAddress: 0,
	blua32StaticLayoutTokenLo: 0,
	blua32StaticLayoutTokenHi: 0,
	blua32DiagnosticDirectoryOffset: 0,
	metadataOffset: 0,
	metadataLength: 0,
};

test('cartridge manifest parses concrete hardware devices in declaration order', () => {
	assert.deepEqual(parseCartManifest({
		title: 'Expansion test',
		hardware: [
			{ type: 'rom' },
			{ type: 'ram', bytes: 0x20000 },
			{ type: 'mailbox' },
		],
	}, 'manifest'), {
		title: 'Expansion test',
		hardware: [
			{ type: 'rom' },
			{ type: 'ram', bytes: 0x20000 },
			{ type: 'mailbox' },
		],
	});
});

test('cartridge manifest requires the current hardware schema', () => {
	assert.throws(
		() => parseCartManifest({ title: 'Missing hardware' }, 'manifest'),
		/manifest\.hardware.*required/,
	);
	assert.throws(
		() => parseCartManifest({
			title: 'Retired board schema',
			cartridge: { board: 'ram_mailbox', ram_bytes: 256 },
		}, 'manifest'),
		/manifest\.cartridge is not part/,
	);
});

test('cartridge manifest rejects invalid concrete device declarations', () => {
	for (const hardware of [
		[{ type: 'rom' }, { type: 'rom' }],
		[{ type: 'rom', bytes: 1 }],
		[{ type: 'ram', bytes: 0 }],
		[{ type: 'ram', bytes: CART_RAM_SIZE + 1 }],
		[{ type: 'ram', bytes: 1 }, { type: 'ram', bytes: 2 }],
		[{ type: 'mailbox' }, { type: 'mailbox' }],
		[{ type: 'mailbox', bytes: 1 }],
		[{ type: 'math' }],
	]) {
		assert.throws(() => parseCartManifest({ hardware }, 'manifest'));
	}
});

test('binary cartridge manifest is decoded through the shared package owner', () => {
	const payload = encodeBinary({
		title: 'Binary expansion',
		hardware: [
			{ type: 'rom' },
			{ type: 'mailbox' },
			{ type: 'ram', bytes: 256 },
		],
	});
	const rom = new Uint8Array(CART_ROM_HEADER_SIZE + payload.byteLength);
	rom.set(payload, CART_ROM_HEADER_SIZE);
	const header: CartRomHeader = {
		...EMPTY_HEADER,
		manifestOffset: CART_ROM_HEADER_SIZE,
		manifestLength: payload.byteLength,
	};
	writeCartRomHeader(rom, header);

	assert.deepEqual(decodeCartManifest(rom, header), {
		title: 'Binary expansion',
		hardware: [
			{ type: 'rom' },
			{ type: 'mailbox' },
			{ type: 'ram', bytes: 256 },
		],
	});
});

test('binary cartridge manifest requires the integer wire tag for RAM bytes', () => {
	const payload = encodeBinary({
		hardware: [{ type: 'ram', bytes: 256.5 }],
	});
	const floatingTagOffset = payload.lastIndexOf(10); // BinTag.F32.
	assert.notEqual(floatingTagOffset, -1);
	new DataView(
		payload.buffer,
		payload.byteOffset + floatingTagOffset + 1,
		4,
	).setFloat32(0, 256, true);
	const header: CartRomHeader = {
		...EMPTY_HEADER,
		manifestOffset: 0,
		manifestLength: payload.byteLength,
	};
	assert.throws(
		() => decodeCartManifest(payload, header),
		/floating-point values are not accepted/,
	);
});
